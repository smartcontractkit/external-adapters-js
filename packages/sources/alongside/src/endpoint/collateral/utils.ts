import { ethers } from 'ethers'
import { bignumber, number } from 'mathjs'
import axios from 'axios'
import abi from './abi.json'
import { makeLogger } from '@chainlink/external-adapter-framework/util/logger'
import { BalanceType } from '.'

const logger = makeLogger('Alongside  collateral calculation logger')

export class Collateral {
  provider: ethers.providers.JsonRpcProvider
  IndexToken: ethers.Contract
  constructor(rpcUrl: string) {
    this.provider = new ethers.providers.JsonRpcProvider(rpcUrl)
    this.IndexToken = new ethers.Contract(
      '0xF17A3fE536F8F7847F1385ec1bC967b2Ca9caE8D',
      abi,
      this.provider,
    )
  }

  getAssetWeights = async () => {
    const input = await this.getAssetWeightsInput()
    return this.calcAssetWeights(
      input.feeWhenMethodologySet,
      input.feeChanges,
      input.from,
      input.to,
      input.initialMethodology,
    )
  }

  getAssetWeightsInput = async () => {
    logger.debug('Getting asset weights input')
    const indexToken = this.IndexToken
    const from = await this.getLatestMethodologySetInfo(indexToken)
    const latestBlock = await this.provider.getBlock('latest')
    const to = {
      timestamp: latestBlock.timestamp,
      blockNumber: latestBlock.number,
    }

    // fee at the time the methodology was set
    const feeWhenMethodologySet =
      (await indexToken.feeRatePerDayScaled({
        blockTag: from.blockNumber,
      })) / Math.pow(10, 20)

    // get all intervening fee changes
    const feeChanges = await this.getFeeChanges(indexToken, from.blockNumber, to.blockNumber)
    logger.debug('Fetching methodology')
    const initialMethodology = await this.fetchMethodology()
    return {
      feeWhenMethodologySet,
      feeChanges,
      from,
      to,
      initialMethodology,
    }
  }

  calcAssetWeights = (
    feeWhenMethodologySet: number,
    feeChanges: any,
    from: { blockNumber?: number; timestamp: number },
    to: { timestamp: number; blockNumber?: number },
    initialMethodology: { [s: string]: unknown } | ArrayLike<unknown>,
  ) => {
    logger.debug('Calculating asset weights')
    const intervals = [
      ...feeChanges,
      // and add the timestamp of the last mint/redeem
      [to.timestamp, NaN],
    ]

    const inflation = this.calcMultiRateInflation(from.timestamp, feeWhenMethodologySet, intervals)

    return Object.fromEntries(
      Object.entries(initialMethodology).map(([symbol, weight]) => [
        symbol,
        (weight as number) * inflation,
      ]),
    )
  }

  calcMultiRateInflation = (
    startTimestamp: number,
    startFee: number,
    intervals: Iterable<number[]>,
  ) => {
    logger.debug('Calculating multi rate inflation')
    let inflation = 1
    let lastTimestamp = startTimestamp
    let lastFeeRate = startFee
    for (const [timestamp, feeRate] of intervals) {
      inflation *= this.calcInflation(lastFeeRate, lastTimestamp, timestamp)
      lastFeeRate = feeRate
      lastTimestamp = timestamp
    }
    return inflation
  }

  calcInflation = (feeRate: number, fromTimestamp: number, toTimestamp: number) => {
    logger.debug('Calculating inflation')
    const bigNumResult = bignumber(1)
      .div(bignumber(1).plus(feeRate))
      .pow(
        bignumber(toTimestamp)
          .sub(bignumber(fromTimestamp))
          .div(60 * 60 * 24),
      )
    return number(bigNumResult)
  }

  getLatestMethodologySetInfo = async (indexToken: ethers.Contract) => {
    logger.debug('Getting latest methodology info')
    const latestBlock = await this.provider.getBlock('latest')
    const logs = await this.provider.getLogs({
      ...indexToken.filters.MethodologySet(),
      fromBlock: latestBlock.number - 300_000,
      toBlock: latestBlock.number,
    })
    // TODO replace me with an assert
    if (logs.length === 0) {
      throw new Error("shouldn't happen")
    }
    const blockNumber = logs[logs.length - 1].blockNumber
    return await this.withTimestamp(blockNumber)
  }

  getFeeChanges = async (indexToken: ethers.Contract, fromBlock: number, toBlock: number) => {
    logger.debug('Getting fee changes')
    return await Promise.all(
      (
        await this.provider.getLogs({
          ...indexToken.filters.FeeRateSet(),
          // TODO test that this won't mess up when they're the same number
          // TODO test we exclude the fee change when
          // it's _before_ the methodology set in the same block
          fromBlock,
          toBlock,
        })
      ).map(async (feeChange) => [
        await this.getTimestamp(feeChange.blockNumber),
        parseInt(feeChange.topics[1]) / Math.pow(10, 20),
      ]),
    )
  }

  getTimestamp = async (blockNumber: number) =>
    (await this.provider.getBlock(blockNumber)).timestamp

  withTimestamp = async (blockNumber: number) => ({
    blockNumber,
    timestamp: await this.getTimestamp(blockNumber),
  })

  fetchMethodology = async () =>
    (
      await this.fetchIPFS(
        await this.IndexToken.methodology({
          blockTag: (await this.provider.getBlock('latest')).number,
        }),
        false,
      )
    ).assets

  fetchIPFS = async (path: string, pinataFirst = true) => {
    try {
      if (pinataFirst) {
        const data = await this.fetchIPFSFromPinata(path)
        return data
      } else {
        const data = await this.fetchIPFSFromInfura(path)
        return data
      }
    } catch (error) {
      logger.error(
        `failed to fetch from ${pinataFirst ? 'Pinata' : 'Infura'}, trying ${
          pinataFirst ? 'Infura' : 'Pinata'
        }"}`,
      )
      if (pinataFirst) {
        const data = await this.fetchIPFSFromInfura(path)
        return data
      } else {
        const data = await this.fetchIPFSFromPinata(path)
        return data
      }
    }
  }

  calcMinCollateral = (
    tradingBalances: BalanceType[],
    vaultBalances: BalanceType[],
    units: { [x: string]: number },
  ): number => {
    logger.warn({
      tradingBalances,
      vaultBalances,
      units,
    })
    let min = Number.MAX_SAFE_INTEGER
    const balances: { [key: string]: number } = {}
    for (const key of Object.keys(units)) {
      for (let i = 0; i < tradingBalances.length; i++) {
        if (tradingBalances[i].symbol.toUpperCase() === key) {
          balances[key] = tradingBalances[i].amount / units[key]
        }
      }
      for (let i = 0; i < vaultBalances.length; i++) {
        if (vaultBalances[i].symbol.toUpperCase() === key) {
          if (Object.keys(balances).includes(key)) {
            balances[key] += vaultBalances[i].amount / units[key]
          } else {
            balances[key] = vaultBalances[i].amount / units[key]
          }
        }
      }
    }
    if (Object.keys(balances).sort().join(',') !== Object.keys(units).sort().join(',')) {
      throw 'Balances not found for all units'
    }
    for (const key of Object.keys(balances)) {
      if (balances[key] < min) {
        min = balances[key]
      }
    }
    return min
  }

  fetchIPFSFromPinata = async (path: string) => {
    logger.debug('Fetching IPFS from Pinata')
    const newPath = path.replace('ipfs://', '')
    const baseURL = 'dxas'
    const ipfsUrl = `${baseURL}/${newPath}`
    const res = await axios.get(ipfsUrl, {
      headers: {
        Accept: 'Accept: text/plain',
      },
    })
    return res?.data
  }

  fetchIPFSFromInfura = async (path: string) => {
    logger.debug('Fetching IPFS from Infura')
    const newPath = path.replace('ipfs://', '')
    const baseURL = 'https://amkt.infura-ipfs.io/ipfs'
    const ipfsUrl = `${baseURL}/${newPath}`

    const res = await axios.get(ipfsUrl, {
      headers: {
        Accept: 'Accept: text/plain',
      },
    })
    return res?.data
  }
}
