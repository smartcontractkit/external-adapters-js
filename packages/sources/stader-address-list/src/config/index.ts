import { AdapterConfig } from '@chainlink/external-adapter-framework/config'

export const config = new AdapterConfig(
  {
    RPC_URL: {
      type: 'string',
      description:
        'The RPC URL to connect to the EVM chain the address manager contract is deployed to.',
      required: true,
    },
    CHAIN_ID: {
      type: 'number',
      description: 'The chain id to connect to',
      default: 1,
    },
    BACKGROUND_EXECUTE_MS: {
      description:
        'The amount of time the background execute should sleep before performing the next request',
      type: 'number',
      default: 10_000,
    },
  },
  {
    envDefaultOverrides: {
      BACKGROUND_EXECUTE_TIMEOUT: 180_000,
    },
  },
)
