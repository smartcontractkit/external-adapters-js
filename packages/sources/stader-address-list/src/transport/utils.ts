import { makeLogger } from '@chainlink/external-adapter-framework/util'
import { BasicAddress } from '../endpoint/address'

const logger = makeLogger('StaderAddressListUtil')

export type NetworkChainMap = {
  [network: string]: {
    [chain: string]: { staderConfig: string }
  }
}

export type ValidatorRegistryResponse = [
  status: number,
  pubkey: string,
  preDepositSignature: string,
  depositSignature: string,
  withdrawVaultAddress: string,
  operatorId: number,
  initialBondEth: number,
  depositTime: number,
  withdrawnTime: number,
]

export const filterDuplicateAddresses = <T extends BasicAddress>(addresses: T[]): T[] => {
  const addressMap: Record<string, T> = {}
  for (const addressObject of addresses) {
    if (addressMap[addressObject.address]) {
      logger.warn(
        { warning: 'Duplicate address detected' },
        `The address "${addressObject.address}" is duplicated in the request and the duplicate has been removed.`,
      )
    } else {
      addressMap[addressObject.address] = addressObject
    }
  }
  return Object.values(addressMap)
}

export async function runAllSequentially<T>({
  count,
  handler,
}: {
  count: number
  handler: (i: number) => Promise<T>
}): Promise<T[]> {
  const results: T[] = []
  for (let i = 0; i < count; i++) {
    results.push(await handler(i))
  }
  return results
}
