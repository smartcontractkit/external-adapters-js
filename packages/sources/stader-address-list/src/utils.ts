import { makeLogger } from '@chainlink/external-adapter-framework/util'

const logger = makeLogger('StaderAddressListUtil')

export type NetworkChainMap = {
  [network: string]: {
    [chain: string]: { staderConfig: string }
  }
}

export type ValidatorRegistryResponse = [
  status: number,
  pubkey: string,
  preDepositSignature: string,
  depositSignature: string,
  withdrawVaultAddress: string,
  operatorId: number,
  initialBondEth: number,
  depositTime: number,
  withdrawnTime: number,
]

export type BasicAddress = {
  address: string
}

export type PoolAddress = BasicAddress & {
  poolId: number
}

export type ValidatorAddress = BasicAddress &
  PoolAddress & {
    network: string
    chainId: string
    withdrawVaultAddress: string
    operatorId: number
    status: number
  }

export interface ResponseSchema {
  Data: {
    stakeManagerAddress?: string
    poolFactoryAddress?: string
    penaltyAddress?: string
    permissionedPoolAddress?: string
    staderConfigAddress?: string
    validatorStatus?: string[]
    socialPoolAddresses: PoolAddress[]
    elRewardAddresses: BasicAddress[]
    confirmations: number
    network: string
    chainId: string
    result: ValidatorAddress[]
  }
  Result: null
}

export const filterDuplicateAddresses = <T extends BasicAddress>(addresses: T[]): T[] => {
  const addressMap: Record<string, T> = {}
  for (const addressObject of addresses) {
    if (addressMap[addressObject.address]) {
      logger.warn(
        { warning: 'Duplicate address detected' },
        `The address "${addressObject.address}" is duplicated in the request and the duplicate has been removed.`,
      )
    } else {
      addressMap[addressObject.address] = addressObject
    }
  }
  return Object.values(addressMap)
}

export async function runAllSequentially<T>({
  count,
  handler,
}: {
  count: number
  handler: (i: number) => Promise<T>
}): Promise<T[]> {
  const results: T[] = []
  for (let i = 0; i < count; i++) {
    results.push(await handler(i))
  }
  return results
}
