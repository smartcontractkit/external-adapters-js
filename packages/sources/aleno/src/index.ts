import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
