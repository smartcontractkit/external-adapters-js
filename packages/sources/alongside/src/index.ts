import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
