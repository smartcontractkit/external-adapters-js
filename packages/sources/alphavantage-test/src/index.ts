import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
