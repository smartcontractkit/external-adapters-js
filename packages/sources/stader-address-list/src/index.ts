import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PoRAdapter } from '@chainlink/external-adapter-framework/adapter/por'
import { config } from './config'
import { addressEndpoint } from './endpoint/address'

export const adapter = new PoRAdapter({
  name: 'STADER_ADDRESS_LIST',
  endpoints: [addressEndpoint],
  defaultEndpoint: addressEndpoint.name,
  config,
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
