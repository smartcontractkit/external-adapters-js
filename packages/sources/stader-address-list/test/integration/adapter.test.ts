import request, { SuperTest, Test } from 'supertest'
import { createAdapter } from './setup'
import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { AddressInfo } from 'net'
import { ValidatorRegistryResponse } from '../../src/utils'

const mockNodeRegistryAddresses: Record<number, string> = {
  1: '0xfb2E1e5854Ba5De7e4611E2352cfe85d91106291',
  2: '0x50297e640b62281b6Dac0d5Aa91848Fb028357Ea',
}

const mockValidatorAddresses: Record<number, ValidatorRegistryResponse[]> = {
  1: [
    [
      4,
      '0x82aec070f2c77f13ef33647c6b6df6bfcaa1cb3607e4530801a753099c4a523005ff701ccb296da9d4488ddd5ac531c3',
      '0xa81f0e3ef17847e61e21a6ac8e2276a7af54df53c2573f653a9a5815f4b81101b1b9fd9607e03c052a39464c524e22ed0670af1ea8b5a0e5a3cfbe3e0caae491a2ab763b61220142b3879f95972f251e9de0efca8f5adeb47b8d68183ac3bc0c',
      '0x9469523ff1410d4417e2da95c6a8b86aa697056425818ce1ba1e631f4e6666326a1567dbaf2fd452337d3def0ec2898a002aba36f8e0bd9964e85b5a3a377caa8555d9f15a8925c594bc422fa2b7aafc27d747f33375151b5ab215c4ad274f7e',
      '0x05ef9b38092ebd70523275f95027709aeac9759c',
      2,
      4000000000000000000,
      1679924532,
      0,
    ],
    [
      4,
      '0x893b284977dbd07276d5e0f3761dddad88aa4e1ed8575541c79a5bb12fdc841dcbe9ecfb424ee7744db4f5e297c0eaf4',
      '0xb2ab614bfebff2ac7ad9ed178b0aa40555539e5106d27cc3d3ad60d9a14a3e2ff75371212a2a205115f991eab25bb2c902b65632b633ca48ba5838e0223331a3c2246cca3fe832eb988de302a10eceffc98a3ad445957bc26bcfc54919cdc727',
      '0x8a10dbc42bb27e0be5d26579e29cddab2eadc1a493879bbc47b78575d17aa6b231bc874d5408cc3408db79034c03c2eb175fbbba0b9cdc6bbf81701aa3e39c9eda2f1293d75e5cbeef704967626748141cdfcc6051cc24e2714595a27606efcf',
      '0xf43d5b8aa1ca682f6fb804899117bfce3a67dbd1',
      2,
      4000000000000000000,
      1679934732,
      0,
    ],
  ],
  2: [
    [
      3,
      '0x999d1e56148505251d7a7682996f1d1a7e5ecf4657dcfa9d6d71b97dc1d8755e1cb4451f57f370201c882126e97ca35e',
      '0x86f71619e07b4423abf7ac8122138bc13fa1271b6fbb3afe6e2bcedc84655683e7a0486c00764c21ceb1b145f2c65d180e0356839c2587edaf10f43368744794fb9e74d4d728b62de3cc145d955ad920281b1176799b2496772aa9552dfab120',
      '0x86c2cd91227ac34fd1a6edde0fd20dbb4ce625f3e0eb47c8f8d0eb62588ddf54255044efb11e0a1177bbbc52ea7d499716b2bb59ae1fde016d7c690fec142649d89215b07d71a3d0d6463042addb9fc29ae35b62434ca9e49c84e3544a3c9b5c',
      '0xe731fd3b95136dac85cfd0c3164707a9e10026f1',
      2,
      4000000000000000000,
      1679934732,
      0,
    ],
    [
      6,
      '0x89cbed0fe0b742c5ad8e9068650dfd6789aa64f2a8e264caf1cbb9e3a50c6eb1ed6b08f2075a384b1666c8e073a0e51d',
      '0x96e61684831b7d18a36ae491f578d1ee20105d0c84704a179b22764b82c80e7eb03ff1bdd2d9d7c82dec679f1f5871e20580c7b8a31ac54f1ed5c7785a41f69e2a6de08d752797c8dec32a3dc37c722f01cc24e6eaf2bb6adc6203e18f97bd48',
      '0xaaec961158b897cb1e2770db067f7e9c564da7c45ab14710b3b5ee4a8fdb82456061517e2ed86b23e2afbd8da3d5fe650d62eaf96a47714cd2bdd138dba6fd9b2425356e0ce625ce6933ec39520b50c90f5caa2fc556687b973248c406f180da',
      '0x2f32b2ed50d03bcf925cd32d870a12d1e33556aa',
      2,
      4000000000000000000,
      1679934732,
      0,
    ],
  ],
}

const mockSocialPoolAddresses: Record<number, string> = {
  1: '0xEB31F35afF9036bcDEC75604D9492aC9c446127F',
  2: '0x481be4fACe9E50bae3a716801F3c8799020325a0',
}

const mockElRewardAddresses: string[] = [
  '0x0D086a2FB3be4251221066a491645273C61A45Fc',
  '0x6D75F292A22fCE61d64448789549ae735eFEC398',
]

jest.mock('ethers', () => {
  const actualModule = jest.requireActual('ethers')
  let getAllActiveValidatorsCallCount = 1
  return {
    ...actualModule,
    ethers: {
      ...actualModule.ethers,
      providers: {
        JsonRpcProvider: function () {
          return {
            getBlockNumber: jest.fn().mockReturnValue(1000),
          }
        },
      },
      Contract: function () {
        return {
          getNodeRegistry: jest.fn().mockImplementation((poolId) => {
            return mockNodeRegistryAddresses[poolId]
          }),
          getPoolIdArray: jest.fn().mockReturnValue([1, 2]),
          getAllActiveValidators: jest.fn().mockImplementation(() => {
            return mockValidatorAddresses[(getAllActiveValidatorsCallCount++ % 2) + 1]
          }),
          getSocializingPoolAddress: jest.fn().mockImplementation((poolId) => {
            return mockSocialPoolAddresses[poolId]
          }),
          getAllSocializingPoolOptOutOperators: jest.fn().mockImplementation(() => {
            return mockElRewardAddresses
          }),
          nextValidatorId: jest.fn().mockReturnValue(3),
          nextOperatorId: jest.fn().mockReturnValue(3),
          getPoolUtils: jest.fn().mockReturnValue('0x9bE5fd05529450c4ABA3c1525b82EF69Af9b3253'),
          getPermissionlessNodeRegistry: jest
            .fn()
            .mockReturnValue('0x28d5FdcE4db361e19FCA420C9b5141bC3DC6aE3B'),
        }
      },
    },
  }
})

describe('execute', () => {
  let fastify: ServerInstance | undefined
  let req: SuperTest<Test>
  let spy: jest.SpyInstance

  beforeAll(async () => {
    const mockDate = new Date('2022-01-01T11:11:11.111Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())
  })

  afterAll((done) => {
    spy.mockRestore()
    fastify?.close(done())
  })

  it('addresses should return success', async () => {
    const data = {
      data: { network: 'ethereum', chainId: 'goerli' },
    }
    process.env['RPC_URL'] = 'http://localhost:9091'
    process.env['BACKGROUND_EXECUTE_MS'] = '0'
    fastify = await expose(createAdapter())
    req = request(`http://localhost:${(fastify?.server.address() as AddressInfo).port}`)
    const mockDate = new Date('2022-08-01T07:14:54.909Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())

    const response = await req
      .post('/')
      .send(data)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
  })
})
