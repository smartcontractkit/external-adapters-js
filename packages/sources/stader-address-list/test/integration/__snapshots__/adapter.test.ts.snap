// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute addresses should return success 1`] = `
{
  "data": {
    "chainId": "goerli",
    "confirmations": 0,
    "elRewardAddresses": [
      {
        "address": "0x0D086a2FB3be4251221066a491645273C61A45Fc",
      },
      {
        "address": "0x6D75F292A22fCE61d64448789549ae735eFEC398",
      },
    ],
    "network": "ethereum",
    "result": [
      {
        "address": "0x999d1e56148505251d7a7682996f1d1a7e5ecf4657dcfa9d6d71b97dc1d8755e1cb4451f57f370201c882126e97ca35e",
        "chainId": "goerli",
        "network": "ethereum",
        "operatorId": 2,
        "poolId": 1,
        "status": 3,
        "withdrawVaultAddress": "0xe731fd3b95136dac85cfd0c3164707a9e10026f1",
      },
      {
        "address": "0x89cbed0fe0b742c5ad8e9068650dfd6789aa64f2a8e264caf1cbb9e3a50c6eb1ed6b08f2075a384b1666c8e073a0e51d",
        "chainId": "goerli",
        "network": "ethereum",
        "operatorId": 2,
        "poolId": 1,
        "status": 6,
        "withdrawVaultAddress": "0x2f32b2ed50d03bcf925cd32d870a12d1e33556aa",
      },
      {
        "address": "0x82aec070f2c77f13ef33647c6b6df6bfcaa1cb3607e4530801a753099c4a523005ff701ccb296da9d4488ddd5ac531c3",
        "chainId": "goerli",
        "network": "ethereum",
        "operatorId": 2,
        "poolId": 2,
        "status": 4,
        "withdrawVaultAddress": "0x05ef9b38092ebd70523275f95027709aeac9759c",
      },
      {
        "address": "0x893b284977dbd07276d5e0f3761dddad88aa4e1ed8575541c79a5bb12fdc841dcbe9ecfb424ee7744db4f5e297c0eaf4",
        "chainId": "goerli",
        "network": "ethereum",
        "operatorId": 2,
        "poolId": 2,
        "status": 4,
        "withdrawVaultAddress": "0xf43d5b8aa1ca682f6fb804899117bfce3a67dbd1",
      },
    ],
    "socialPoolAddresses": [
      {
        "address": "0xEB31F35afF9036bcDEC75604D9492aC9c446127F",
        "poolId": 1,
      },
      {
        "address": "0x481be4fACe9E50bae3a716801F3c8799020325a0",
        "poolId": 2,
      },
    ],
    "validatorStatus": [],
  },
  "result": null,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1659338094909,
    "providerDataRequestedUnixMs": 1659338094909,
  },
}
`;
